### PR TITLE
GitHub Action options to invoke linters through Darker

### LINKED ISSUE #281: Add option to add linters (flake8, mypy, etc.) in the GH Action
There is no other way to add the supported linters other than creating separate steps to install.

### PR DESCRIPTION
**Pull request by akaihola**
Fixes #281.

Follow-up issue #337 will allow pinning linter packages.

### CODE REVIEW

**Comment by akaihola** [comment]
@roniemartinez, could you review this PR? If you are willing to accept the invitation I've sent to you to become a collaborator on this project, I can assign you as a reviewer.

File: action/tests/test_main.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    ),
+)
+def test_runs_darker(tmp_path, env, expect):
+    """Configuration translates correctly into a Darker command line"""
+    with patch_main(tmp_path, env) as main_patch, pytest.raises(SysExitCalled):
+
+        run_module("main")
+
+    darker = str(tmp_path / ".darker-env" / BIN / "darker")
+    # We can change `c[0][0][0]` to `c.args[0][0]` after droppin support for Python 3.7.
```
Comment (roniemartinez) [inline action/tests/test_main.py:205]:
```suggestion
    # We can change `c[0][0][0]` to `c.args[0][0]` after dropping support for Python 3.7.
```

**Review by roniemartinez — APPROVED** [review]
Awesome job! Thanks for implementing this feature. ⭐

**Comment by cclauss** [comment]
bandit --recursive --skip B101 .  # B101 is assert statements which are in ./action/tests/test_main.py

**Comment by akaihola** [comment]
Hm, [Bandit 1.7.4](https://pypi.org/project/bandit/1.7.4/) was released on March 4th, 2022, and [`bandit-report-artifacts/requirements.txt`](https://github.com/ioggstream/bandit-report-artifacts/blob/d706d74be4004d4300a56feca4b6ca98d70211ac/requirements.txt) doesn't pin the Bandit version. Maybe the 1.7.4 version changed Bandit's behavior. I'll look into this.

**Comment by cclauss** [comment]
Might also get some hints from https://github.com/PyCQA/bandit-action

**Comment by akaihola** [comment]
I tried to see if Bandit 1.7.2 vs 1.7.4 would behave differently, but I don't see any differences. I believe the command lines I use are the same which [`bandit-report-artifacts/entrypoint.sh`](https://github.com/ioggstream/bandit-report-artifacts/blob/d706d74be4004d4300a56feca4b6ca98d70211ac/entrypoint.sh) executes (apart from the `-o <filename>.txt` option to write results into a text file):

```bash
$ bandit --version
bandit 1.7.2
  python version = 3.10.0 (default, Oct  4 2021, 17:55:55) [GCC 10.3.0]
```

<details>
<summary><code>$ bandit -c .bandit.code.yaml -r . -f txt</code></summary>

```bash
[main]	INFO	profile include tests: None
[main]	INFO	profile exclude tests: None
[main]	INFO	cli include tests: None
[main]	INFO	cli exclude tests: None
[main]	INFO	using config: .bandit.code.yaml
[main]	INFO	running on Python 3.10.0
Run started:2022-03-23 18:31:07.064305

Test results:
	No issues identified.

Code scanned:
	Total lines of code: 2779
	Total lines skipped (#nosec): 9

Run metrics:
	Total issues (by severity):
		Undefined: 0.0
		Low: 0.0
		Medium: 0.0
		High: 0.0
	Total issues (by confidence):
		Undefined: 0.0
		Low: 0.0
		Medium: 0.0
		High: 0.0
Files skipped (0):
```
</details>

<details>
<summary><code>$ bandit -c .bandit.tests.yaml -r src/darker/tests -f txt</code></summary>

```bash
[main]	INFO	profile include tests: None
[main]	INFO	profile exclude tests: B101
[main]	INFO	cli include tests: None
[main]	INFO	cli exclude tests: None
[main]	INFO	using config: .bandit.tests.yaml
[main]	INFO	running on Python 3.10.0
Run started:2022-03-23 18:30:28.711222

Test results:
	No issues identified.

Code scanned:
	Total lines of code: 4275
	Total lines skipped (#nosec): 4

Run metrics:
	Total issues (by severity):
		Undefined: 0.0
		Low: 0.0
		Medium: 0.0
		High: 0.0
	Total issues (by confidence):
		Undefined: 0.0
		Low: 0.0
		Medium: 0.0
		High: 0.0
Files skipped (0):
```
</details>

<details>
<summary><code>$ bandit -c .bandit.tests.yaml -r action/tests -f txt</code></summary>

```bash
[main]	INFO	profile include tests: None
[main]	INFO	profile exclude tests: B101
[main]	INFO	cli include tests: None
[main]	INFO	cli exclude tests: None
[main]	INFO	using config: .bandit.tests.yaml
[main]	INFO	running on Python 3.10.0
Run started:2022-03-23 18:30:13.909150

Test results:
	No issues identified.

Code scanned:
	Total lines of code: 190
	Total lines skipped (#nosec): 1

Run metrics:
	Total issues (by severity):
		Undefined: 0.0
		Low: 0.0
		Medium: 0.0
		High: 0.0
	Total issues (by confidence):
		Undefined: 0.0
		Low: 0.0
		Medium: 0.0
		High: 0.0
Files skipped (0):
```
</details>

```bash
$ pip install -U bandit
Collecting bandit
  Using cached bandit-1.7.4-py3-none-any.whl (118 kB)
Installing collected packages: bandit
  Attempting uninstall: bandit
    Found existing installation: bandit 1.7.2
    Uninstalling bandit-1.7.2:
      Successfully uninstalled bandit-1.7.2
Successfully installed bandit-1.7.4
```

<details>
<summary><code>$ bandit -c .bandit.code.yaml -r . -f txt</code></summary>

```bash
[main]	INFO	profile include tests: None
[main]	INFO	profile exclude tests: None
[main]	INFO	cli include tests: None
[main]	INFO	cli exclude tests: None
[main]	INFO	using config: .bandit.code.yaml
[main]	INFO	running on Python 3.10.0
Run started:2022-03-23 18:31:17.317269

Test results:
	No issues identified.

Code scanned:
	Total lines of code: 2779
	Total lines skipped (#nosec): 9
	Total potential issues skipped due to specifically being disabled (e.g., #nosec BXXX): 0

Run metrics:
	Total issues (by severity):
		Undefined: 0
		Low: 0
		Medium: 0
		High: 0
	Total issues (by confidence):
		Undefined: 0
		Low: 0
		Medium: 0
		High: 0
Files skipped (0):
```
</details>

<details>
<summary><code>$ bandit -c .bandit.tests.yaml -r src/darker/tests -f txt</code></summary>

```bash
[main]	INFO	profile include tests: None
[main]	INFO	profile exclude tests: B101
[main]	INFO	cli include tests: None
[main]	INFO	cli exclude tests: None
[main]	INFO	using config: .bandit.tests.yaml
[main]	INFO	running on Python 3.10.0
Run started:2022-03-23 18:31:21.972722

Test results:
	No issues identified.

Code scanned:
	Total lines of code: 4275
	Total lines skipped (#nosec): 4
	Total potential issues skipped due to specifically being disabled (e.g., #nosec BXXX): 0

Run metrics:
	Total issues (by severity):
		Undefined: 0
		Low: 0
		Medium: 0
		High: 0
	Total issues (by confidence):
		Undefined: 0
		Low: 0
		Medium: 0
		High: 0
Files skipped (0):
```
</details>

<details>
<summary><code>$ bandit -c .bandit.tests.yaml -r action/tests -f txt</code></summary>

```bash
[main]	INFO	profile include tests: None
[main]	INFO	profile exclude tests: B101
[main]	INFO	cli include tests: None
[main]	INFO	cli exclude tests: None
[main]	INFO	using config: .bandit.tests.yaml
[main]	INFO	running on Python 3.10.0
Run started:2022-03-23 18:31:25.908424

Test results:
	No issues identified.

Code scanned:
	Total lines of code: 190
	Total lines skipped (#nosec): 1
	Total potential issues skipped due to specifically being disabled (e.g., #nosec BXXX): 0

Run metrics:
	Total issues (by severity):
		Undefined: 0
		Low: 0
		Medium: 0
		High: 0
	Total issues (by confidence):
		Undefined: 0
		Low: 0
		Medium: 0
		High: 0
Files skipped (0):
```
</details>

**Comment by akaihola** [comment]
I also tried using Bandit 1.7.3, and the results were still the same. So this still remains a mystery.

One thing I now realized is that each of the three Bandit steps writes results into the same file (see [`bandit-report-artifacts/entrypoint.sh:12`](https://github.com/ioggstream/bandit-report-artifacts/blob/d706d74be4004d4300a56feca4b6ca98d70211ac/entrypoint.sh#L12)):
```bash
-o "${GITHUB_WORKSPACE}/output/security_report.txt"
```

I'll see if separating the runs into different builds instead of steps in the same build would help.

**Comment by akaihola** [comment]
Separate jobs didn't help.

This looks really strange. In this screenshot, you can see the three Bandit steps succeed (`bandit-code`, `bandit-tests` and `bandit-action-tests`). But before them as part of the `lintAllTheThings` job from `.github/workflows/yaml-lint.yml` we can see `x Bandit comments` which is the step causing the failure.

image

**Comment by akaihola** [comment]
It seems that the title `Bandit comments` comes from [bandit-report-artifacts/main.py:99](https://github.com/ioggstream/bandit-report-artifacts/blob/d706d74be4004d4300a56feca4b6ca98d70211ac/main.py#L99) in `bandit_run_check()` which is [called](https://github.com/ioggstream/bandit-report-artifacts/blob/d706d74be4004d4300a56feca4b6ca98d70211ac/main.py#L142-L144) from the main program.

The main program then [calls `gh()`](https://github.com/ioggstream/bandit-report-artifacts/blob/d706d74be4004d4300a56feca4b6ca98d70211ac/main.py#L145-L150) with the option `data='{"name": "Bandit comments", ...}'`. `gh()` in turn calls the `/check-runs` GitHub API endpoint with that JSON as the body. The annotations we see in the screenshot above come from `{"output": {"annotations": ...}, ...}` in that same JSON.

So how does all this end up shown in the output from the `Yaml Lint` workflow which doesn't even call the Bandit action?

**Comment by akaihola** [comment]
I think I understand what's happening with Bandit.

In [entrypoint.sh](https://github.com/ioggstream/bandit-report-artifacts/blob/d706d74be4004d4300a56feca4b6ca98d70211ac/entrypoint.sh), `bandit` is called _twice_.

The [first call](https://github.com/ioggstream/bandit-report-artifacts/blob/d706d74be4004d4300a56feca4b6ca98d70211ac/entrypoint.sh#L12) happens directly and uses the configuration file specified in the workflow. Note that `${BANDIT_CONFIG}` is passed as the first argument:
```bash
bandit ${BANDIT_CONFIG} \
  -r "${INPUT_PROJECT_PATH}" \
  -o "${GITHUB_WORKSPACE}/output/security_report.txt" \
  -f 'txt'
```
The second call happens inside [main.py](https://github.com/ioggstream/bandit-report-artifacts/blob/d706d74be4004d4300a56feca4b6ca98d70211ac/main.py) which is [called](https://github.com/ioggstream/bandit-report-artifacts/blob/d706d74be4004d4300a56feca4b6ca98d70211ac/entrypoint.sh#L15) right after the previous call, and notably doesn't pass on the value for `${BANDIT_CONFIG}`:
```bash
GITHUB_TOKEN=$INPUT_REPO_TOKEN python /main.py -r $INPUT_PROJECT_PATH
```
The main program in main.py [calls `run_bandit()`](https://github.com/ioggstream/bandit-report-artifacts/blob/d706d74be4004d4300a56feca4b6ca98d70211ac/main.py#L140), passing on the above command line options.
The value of `argv[1:]` is `["-r", p]` where `p` comes from the `$INPUT_PROJECT_PATH` environment variable:
```python
    bandit_results = run_bandit(argv[1:], env={"PATH": environ["PATH"]})
```
In [`main.run_bandit()`](https://github.com/ioggstream/bandit-report-artifacts/blob/d706d74be4004d4300a56feca4b6ca98d70211ac/main.py#L34-L43), `bandit` is [called](https://github.com/ioggstream/bandit-report-artifacts/blob/d706d74be4004d4300a56feca4b6ca98d70211ac/main.py#L36-L39) with the additional `-f json` option but without the configuration file argument:
```python
    my_args = ["bandit", "-f", "json"] + args
    out = run(  # nosec - this input cannot execute different commands.
        my_args, shell=False, capture_output=True, env=env or {"PATH": environ["PATH"]},
    )
```

So I guess the purpose of the first call in `entrypoint.sh` is to get a plaintext report and the status code from Bandit. In the second call from `main.py`, check annotations are collected as JSON so they can be passed on to GitHub Actions.

Unfortunately the two calls use different configurations in our case, so the latter call fails to apply `skips: B101` from [`.bandit.tests.yaml`](https://github.com/akaihola/darker/blob/c62fde8e1bdaa335de9e629d5d0e5cf68fabde2f/.bandit.tests.yaml) when run on test directories.

@ioggstream, could you verify that I've understood the problem correctly? Would you be willing to accept a patch to [bandit-report-artifacts](https://github.com/ioggstream/bandit-report-artifacts) which makes `main.py` apply the configuration to the `bandit` call as well?

**Comment by akaihola** [comment]
I think I've fixed the `bandit` issue by using my own branch [akaihola/bandit-report-artifacts@use-config](https://github.com/akaihola/bandit-report-artifacts/commits/use-config) (see ioggstream/bandit-report-artifacts#8).

I'll move this change into its own branch since it's unrelated to this PR.

**Comment by ioggstream** [comment]
@akaihola merged, thx++
